### PR TITLE
[MRG+2] adding warning for rank-deficient cov matrix with no regularization

### PR DESCRIPTION
**Pull request by britta-wstnr**
adding warning if covariance matrix is rank-deficient but no regularization is provided

### CODE REVIEW

File: mne/beamformer/_lcmv.py [hunk 1]
```diff
@@ -17,13 +17,18 @@
...
+from ..utils import logger, verbose, warn, estimate_rank
 from .. import Epochs
 from ..externals import six
 
 
 def _reg_pinv(x, reg):
     """Compute a regularized pseudoinverse of a square array."""
+    # Throw a warning if cov matrix is rank-deficient but no regularization parameter is set
+    covrank = estimate_rank(x, tol='auto', norm=False, return_singular=False)
+    if(covrank < x.shape[0] and reg == 0):
```
Comment (agramfort) [inline mne/beamformer/_lcmv.py]: if(covrank < x.shape[0] and reg == 0):
->
if covrank < x.shape[0] and reg == 0.:

File: mne/beamformer/_lcmv.py [hunk 2]
```diff
@@ -17,13 +17,18 @@
...
 from ..source_space import label_src_vertno_sel
-from ..utils import logger, verbose, warn
+from ..utils import logger, verbose, warn, estimate_rank
 from .. import Epochs
 from ..externals import six
 
 
 def _reg_pinv(x, reg):
     """Compute a regularized pseudoinverse of a square array."""
+    # Throw a warning if cov matrix is rank-deficient but no regularization parameter is set
```
Comment (agramfort) [inline mne/beamformer/_lcmv.py]: no need for this comment. Warnign is explicit enough

**Comment by agramfort** [comment]
@britta-wstnr push directly to update your PR.

File: mne/beamformer/_lcmv.py (hunk 2 above)
Comment (britta-wstnr) [inline mne/beamformer/_lcmv.py]: thank you for your review!

**Comment by agramfort** [comment]
@britta-wstnr can you add a test in test_lcmv.py that just test this if statement by just calling _reg_pinv with a rank deficient array?

**Comment by britta-wstnr** [comment]
@agramfort should I define a "test_reg_pinv" (in test_lcmv.py) or should I rather add it to test_lcmv() ?

**Comment by agramfort** [comment]
Whatever makes testing the fastest to run

File: mne/beamformer/_lcmv.py [hunk 3]
```diff
@@ -17,13 +17,17 @@
...
 from ..source_space import label_src_vertno_sel
-from ..utils import logger, verbose, warn
+from ..utils import logger, verbose, warn, estimate_rank
 from .. import Epochs
 from ..externals import six
 
 
 def _reg_pinv(x, reg):
     """Compute a regularized pseudoinverse of a square array."""
+    covrank = estimate_rank(x, tol='auto', norm=False, return_singular=False)
```
Comment (agramfort) [inline mne/beamformer/_lcmv.py]: I realize that this computation is only necessary when reg=0.

you should do:

```
if reg == 0:
    covrank = estimate_rank(x, tol='auto', norm=False, return_singular=False)
    if covrank < x.shape[0]:
            warn('Covariance matrix is rank-deficient, but no regularization is done.')
```

**Comment by agramfort** [comment]
@britta-wstnr besides my nitpick please check why travis is not happy.

thx

**Comment by dengemann** [comment]
The decrease in coverage suggests that the added code increases the complexity of its usage. 

```
if reg == 0:
    covrank = estimate_rank(x, tol='auto', norm=False,
                                                   return_singular=False)
    if covrank < x.shape[0]:
``` 

This is a double conditional.
Maybe we could try to test more cases here. So far we check for one situation it seems.

**Comment by agramfort** [comment]
I think the codecov complaint is not legit.

+1 for MRG on my end.

**Comment by Eric89GXL** [comment]
> The decrease in coverage suggests that the added code increases complexity.

I think that codecov is wrong (it is from time to time) -- the test covers those lines.

LGTM +1 for merge, @dengemann feel free to merge once you're happy.

**Pull request merged**